src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1387 @@
+        for idx, decoder_layer in enumerate(self.layers):
+            if self.output_hidden_states:
+                all_hidden_states += (hidden_states,)
+            layer_state = past_key_values[idx] if past_key_values is not None else None
+            hidden_states, layer_self_attn, layer_past = decoder_layer(
+                hidden_states,
+                encoder_hidden_states,
+                encoder_attn_mask=encoder_padding_mask,
+                layer_state=layer_state,
+                self_attn_mask=self_attn_mask,
+                output_attentions=output_attentions,
+                ngram_mask_matrix=ngram_mask_matrix,
+                i_buckets_main_stream=i_buckets_main_stream,
+                i_bucket_relative_stream=i_bucket_relative_stream,
+                real_positions=real_positions
+            )
+            if use_cache:
+                next_decoder_cache.append(layer_past.copy())
+            if output_attentions:
+                all_self_attns += (layer_self_attn,)
+        hidden_states_list = hidden_states.transpose(0, 1).chunk(1 + self.ngram, 1)
+        encoder_hidden_states = encoder_hidden_states.transpose(0, 1)
+        if use_cache:
+            next_cache = next_decoder_cache
+        else:
+            next_cache = None
+        if not return_dict:
+            return hidden_states_list, next_cache, all_hidden_states, list(all_self_attns)
+        return BaseModelOutputWithPast(
+            last_hidden_state=hidden_states_list, past_key_values=next_cache, hidden_states=all_hidden_states, attentions=all_self_attns
+        )
+
+def fill_with_neg_inf(t):
+    """FP16-compatible function that fills a input_ids with -inf."""
+    return t.float().fill_(float("-inf")).type_as(t)
+
+def _filter_out_falsey_values(tup) -> Tuple:
+    """Remove entries that are None or [] from an iterable."""
+    return tuple(x for x in tup if isinstance(x, torch.Tensor) or x)
+
+@add_start_docstrings(
+    "The bare ProphetNet Model transformer outputting raw hidden-states without any specific head on top.",
+    PROPHETNET_START_DOCSTRING,
+    PROPHETNET_INPUTS_DOCSTRING,
+)
+class ProphetNetModel(ProphetNetPreTrainedModel):
+    def __init__(self, config):
+        super().__init__(config)
+
+        padding_idx, vocab_size, dim_size = config.pad_token_id, config.vocab_size, config.hidden_size
+        self.embed_tokens = nn.Embedding(vocab_size, dim_size, padding_idx=padding_idx)
+        self.encoder = ProphetNetEncoder(config, self.embed_tokens)
+        self.decoder = ProphetNetDecoder(config, self.embed_tokens)
+
+    def get_input_embeddings(self):
+        return self.embed_tokens
+
+    def set_input_embeddings(self, value):
+        self.embed_tokens = value
+        self.encoder.embed_tokens = self.embed_tokens
+        self.decoder.embed_tokens = self.embed_tokens
+
+    def get_output_embeddings(self):
+        return _make_linear_from_emb(self.embed_tokens)  # make it on the fly
```

Comment:
Not to big of a problem for now, but this should be done at init with `tie_weights` function